lib/services/nodes-api-service.js

```diff
@@ -0,0 +1,96 @@
+// Copyright 2015, EMC, Inc.
+
+'use strict';
+
+var di = require('di');
+
+module.exports = nodeApiServiceFactory;
+di.annotate(nodeApiServiceFactory, new di.Provide('Http.Services.Api.Nodes'));
+di.annotate(nodeApiServiceFactory,
+    new di.Inject(
+        'Protocol.TaskGraphRunner',
+        'Services.Waterline',
+        'Errors',
+        'Logger',
+        '_',
+        'Promise'
+    )
+);
+function nodeApiServiceFactory(
+    taskGraphProtocol,
+    waterline,
+    Errors,
+    Logger,
+    _,
+    Promise
+) {
+    var logger = Logger.initialize(nodeApiServiceFactory);
+
+    function NodeApiService() {
+    }
+
+    NodeApiService.prototype.removeNode = function(node) {
+        return taskGraphProtocol.getActiveTaskGraph( { target: node.id })
+        .then(function (graph) {
+            if (graph) {
+                throw new Errors.BadRequestError('Could not remove node ' + node.id +
+                    ', active workflow is running');
+            }
+        })
+        .then(function () {
+            return Promise.settle([
+                //lookups should be destoryed here, only clear node field
+                //as a workaround until the issue that when lookups are cleared
+                //it cannot be updated timely in nodes' next bootup is fixed
+                waterline.lookups.update({ node: node.id },{ node: '' } ),
+                waterline.nodes.destroy({ id: node.id }),
+                waterline.catalogs.destroy({ node: node.id }),
+                waterline.workitems.destroy({ node: node.id })
+                ]);
+        })
+        .then(function () {
+            // Find the enclosure nodes who enclose this compute node
+            if (_.has(node, 'relations')) {
+                var relation = _.find(node.relations, { relationType: 'enclosedBy' });
+                if (relation &&  _.has(relation, 'targets') ) {
+                    return Promise.map(relation.targets, function (enclNodeId) {
+                        return waterline.nodes.needByIdentifier(enclNodeId)
+                        .catch(function (err) {
+                            logger.error("Error Getting Enclosure Node", { error: err });
+                            // If don't find enclosure node, continue to remove this node
+                            return;
+                        });
+                    });
+                }
+            }
+        })
+        .then(function (enclNodes) {
```

Comment:
I think it would be better to make the code in these promise blocks their own named functions, that way we can more easily test their functionality individually, and also have a more easily readable promise chain.